src/main/scala/org/broadinstitute/dsde/workbench/leonardo/dao/GoogleDataprocDAO.scala

```diff
@@ -1,79 +1,244 @@
 
   private lazy val dataproc = {
-    new Dataproc.Builder(GoogleNetHttpTransport.newTrustedTransport,
-      JacksonFactory.getDefaultInstance, getDataProcServiceAccountCredential)
+    new Dataproc.Builder(httpTransport, jsonFactory, getServiceAccountCredential(cloudPlatformScopes))
       .setApplicationName("dataproc").build()
   }
-  private def getDataProcServiceAccountCredential: Credential = {
+
+  private lazy val storage = {
+    new Storage.Builder(httpTransport, jsonFactory, getServiceAccountCredential(storageScopes))
+      .setApplicationName("storage").build()
+  }
+
+
+  private def getServiceAccountCredential(scopes: List[String]): Credential = {
     new GoogleCredential.Builder()
       .setTransport(httpTransport)
       .setJsonFactory(jsonFactory)
       .setServiceAccountId(dataprocConfig.serviceAccount)
-      .setServiceAccountScopes(cloudPlatformScopes.asJava)
-      .setServiceAccountPrivateKeyFromPemFile(new java.io.File(dataprocConfig.serviceAccountPemPath))
+      .setServiceAccountScopes(scopes.asJava)
+      .setServiceAccountPrivateKeyFromPemFile(serviceAccountPemFile)
       .build()
   }
 
+  /* Creates a cluster in the given google project:
+       - Add a firewall rule to the user's google project if it doesn't exist, so we can access the cluster
+       - Create the initialization bucket for the cluster in the leo google project
+       - Upload all the necessary initialization files to the bucket
+       - Create the cluster in the google project
+     Currently, the bucketPath of the clusterRequest is not used - it will be used later as a place to store notebook results */
+  def createCluster(googleProject: GoogleProject, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
+    // Create the firewall rule in the google project if it doesn't already exist, so we can access the cluster
+    updateFirewallRules(googleProject)
 
-  def createCluster(googleProject: String, clusterName: String, clusterRequest: ClusterRequest)(implicit executionContext: ExecutionContext): Future[ClusterResponse] = {
-    val op = build(googleProject, clusterName, clusterRequest)
-    op.map{op =>
-      val metadata = op.getMetadata
-      ClusterResponse(clusterName, googleProject, metadata.get("clusterUuid").toString, metadata.get("status").toString, metadata.get("description").toString, op.getName)}
+    val bucketName = s"${clusterName}-${UUID.randomUUID.toString}"
+    // Create the bucket in leo's google bucket and populate with initialization files
+    val bucketResponse = initializeBucket(dataprocConfig.leoGoogleBucket, clusterName, bucketName)
+    // Once the bucket is ready, build the cluster
+    bucketResponse.flatMap[ClusterResponse] { _ =>
+      val operation = build(googleProject, clusterName, clusterRequest, bucketName)
```

Comment:
Also I get confused every time I see the `build()` method. Maybe we could name it more descriptively, like `buildCluster`?